app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsFragment.kt

```diff
@@ -0,0 +1,549 @@
+            binding.progressBar.hide()
+
+            if (adapter.itemCount == 0) {
+                when (loadState.refresh) {
+                    is LoadState.NotLoading -> {
+                        if (loadState.append is LoadState.NotLoading && loadState.source.refresh is LoadState.NotLoading) {
+                            binding.statusView.show()
+                            binding.statusView.setup(R.drawable.elephant_friend_empty, R.string.message_empty)
+                        }
+                    }
+                    is LoadState.Error -> {
+                        binding.statusView.show()
+                        binding.statusView.setup((loadState.refresh as LoadState.Error).error) { onRefresh() }
+                    }
+                    is LoadState.Loading -> {
+                        binding.progressBar.show()
+                    }
+                }
+            }
+        }
+
+        adapter.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
+            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
+                if (positionStart == 0 && adapter.itemCount != itemCount) {
+                    binding.recyclerView.post {
+                        if (getView() != null) {
+                            binding.recyclerView.scrollBy(
+                                0,
+                                Utils.dpToPx(requireContext(), -30)
+                            )
+                        }
+                    }
+                }
+                if (readingOrder == ReadingOrder.OLDEST_FIRST) {
+                    updateReadingPositionForOldestFirst()
+                }
+            }
+        })
+
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewModel.notifications.collectLatest { pagingData ->
+                adapter.submitData(pagingData)
+            }
+        }
+
+        viewLifecycleOwner.lifecycleScope.launch {
+            eventHub.events.collect { event ->
+                if (event is PreferenceChangedEvent) {
+                    onPreferenceChanged(event.preferenceKey)
+                }
+            }
+        }
+    }
+
+    override fun onRefresh() {
+        adapter.refresh()
+    }
+
+    override fun onViewAccount(id: String) {
+        super.viewAccount(id)
+    }
+
+    override fun onMute(mute: Boolean, id: String, position: Int, notifications: Boolean) {
+        // not needed, muting via the more menu on statuses is handled in SFragment
+    }
+
+    override fun onBlock(block: Boolean, id: String, position: Int) {
+        // not needed, blocking via the more menu on statuses is handled in SFragment
+    }
+
+    override fun onRespondToFollowRequest(accept: Boolean, id: String, position: Int) {
+        val notification = adapter.peek(position) ?: return
+        val refreshAdapterFlow = viewModel.respondToFollowRequest(accept, accountId = id, notificationId = notification.id)
+        lifecycleScope.launch {
```

Comment:
Yeah that sucks, it would be very nice if I could trigger the refresh from the viewmodel directly, but invalidating the paging source as the documentation suggests does not do anything :/
I'll leave it like this, its not super important to have the notifications refresh.